docs/design/2024-08-27-traffic-replay.md

```diff
@@ -0,0 +1,231 @@
+2. (Optional) The user synchronizes data from the production TiDB cluster to the new cluster and then executes `ANALYZE` commands to refresh the statistics.
+3. The user turns on traffic capture on the production TiDB cluster and TiProxy stores traffic to the specified directory. The username and password should be specified.
+4. The user turns on traffic replay on the testing TiDB cluster and TiProxy replays traffic and generates a report.
+5. The user checks the report to see if the result is expected.
+
+<img src="./imgs/traffic-offline.png" alt="vip architecture" width="600">
+
+### Traffic Capture
+
+TiProxy stores these fields into traffic files:
+- The start timestamp of this command.
+- The TiProxy connection ID of this command.
+- The binary format of this command.
+- The first packet header, packet number, size, and duration of the response.
+
+Most applications use persistent connections, so TiProxy should also capture existing connections. If the existing sessions are in transactions, TiProxy waits for the end of the transaction.
+
+Existing sessions may have states like prepared statements. For these sessions, TiProxy queries the session states by `SHOW SESSION_STATES` and stores the states as `SET SESSION_STATES` statements in the traffic files.
+
+`SHOW SESSION_STATES` shows all session variables, regardless of whether they have been set by the clients or not, which may miss some problems caused by changes of default session variable values. For example, a session variable has changed its default value in the new version, but TiProxy migrates the old value to the testing cluster to overwrite the default value and reports no errors. When the user migrates applications to the new version, he may find some compatibility breakers caused by the new default value.
+
+### Traffic Storage
+
+To support adding more fields in the future, the traffic file format is similar to slow logs:
+
+```
+# Time: 2024-01-19T16:29:48.141142Z
+# Query_time: 0.01
+# Conn_ID: 123
+# Cmd_type: 3
+# Rows_sent: 1
+# Result_size: 100
+# Payload_len: 8
+SELECT 1
+```
+
+To replay the statements in the same order, the statements should be ordered by the start time. So the field `Time` indicates the start time instead of the end time. To order by the start time, one statement must stay in memory until all the previous statements are finished. If a statement runs too long, it's logged without query result information, so the replay skips checking the result and duration.
+
+It's trivial to lose traffic during failure, so TiProxy doesn't need to flush the traffic synchronously. To reduce the effect on the production cluster, TiProxy flushes the traffic in batch asynchronously with double buffering. Similarly, TiProxy loads traffic with double buffering during replaying. If the IO is too slow and the buffer accumulates too much, the capture should stop immediately and the status should be reported in UI and Grafana.
```

Comment:
Could we estimate the performance impact of captures/replays to TiProxy?

Reply:
Yes, latency regression < 1% and throughput < 5%. But I think it should be specified in the spec, not the design.